### PR TITLE
EVM-360 E2E test for tx which contains block functions

### PR DESCRIPTION
**Pull request by igorcrevar**
# Description

E2E for EVM-360 task 

Code of smart contract:
```
// SPDX-License-Identifier: MIT
pragma solidity 0.8.17;

contract Test {
    uint[] public data;
    address public coinbase;
    // bytes32 public hash;

    function init() external {
        data.push(block.difficulty);
        data.push(block.number);
        data.push(block.timestamp);        
        data.push(block.gaslimit);
        data.push(block.chainid);
        coinbase = block.coinbase;
        // data.push(block.basefee);
        // hash = blockhash(block.number);
    }
}
```

- `hash = blockhash(block.number);` in order for this instruction to work we need big refactoring of block builder and some changes in fsm
- `data.push(block.basefee);` this will work after London fork is implemented ([EIP-3198](https://eips.ethereum.org/EIPS/eip-3198) and [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559))

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [ ] I have tested this code manually

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: e2e-polybft/txpool_test.go [hunk 1]
```diff
@@ -159,6 +161,36 @@ func TestE2E_TxPool_Transfer_Linear(t *testing.T) {
...
+	require.NoError(t, err)
+
+	cluster := framework.NewTestCluster(t, 4,
+		framework.WithPremine(types.Address(sidechainKey.Address())),
+	)
+	defer cluster.Stop()
+
+	require.NoError(t, cluster.WaitForBlock(5, 30*time.Second))
+
+	l2TxRelayer, err := txrelayer.NewTxRelayer(txrelayer.WithIPAddress(cluster.Servers[0].JSONRPCAddr()))
```
Comment (ferranbt) [inline e2e-polybft/txpool_test.go]:
```suggestion
	relayer, err := txrelayer.NewTxRelayer(txrelayer.WithIPAddress(cluster.Servers[0].JSONRPCAddr()))
```

File: e2e-polybft/txpool_test.go [hunk 2]
```diff
@@ -159,6 +161,36 @@ func TestE2E_TxPool_Transfer_Linear(t *testing.T) {
...
...0a01b03191641179055565b005b610109610104366004610138565b610117565b60405190815260200161006d565b6000818154811061012757600080fd5b600091825260209091200154905081565b60006020828403121561014a57600080fd5b503591905056fea26469706673582212207602674d664de555f28f6e6014c9c71da8bd936e8a3d4761348319b1d53ebbbb64736f6c63430008110033\",\n  \"deployedBytecode\": \"0x608060405234801561001057600080fd5b50600436106100415760003560e01c8063a6ae0aac14610046578063e1c7392a14610076578063f0ba8440146100f6575b600080fd5b600154610059906001600160a01b031681565b6040516001600160a01b0390911681526020015b60405180910390f35b6100f46000805460018181018355828052447f290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e56392830155825480820184554390830155825480820184554290830155825480820184554590830155825480820190935546929091019190915580546001600160a01b03191641179055565b005b610109610104366004610138565b610117565b60405190815260200161006d565b6000818154811061012757600080fd5b600091825260209091200154905081565b60006020828403121561014a57600080fd5b503591905056fea26469706673582212207602674d664de555f28f6e6014c9c71da8bd936e8a3d4761348319b1d53ebbbb64736f6c63430008110033\",\n  \"linkReferences\": {},\n  \"deployedLinkReferences\": {}\n}\n"
```
Comment (ferranbt) [inline e2e-polybft/txpool_test.go:165]: Could we have some process to generate this artifact from the smart contract and also commit the sm to the repo? Otherwise, it is not straightforward what to do when we need to change the contract.

Comment (igorcrevar) [reply e2e-polybft/txpool_test.go:165]: Sadly `cc := &testutil.Contract{}` -> `Compile` does not work, at least for all variables :( Do you have any other idea?

**Review by b00ris — APPROVED** [review]
Looks good to me

File: e2e-polybft/txpool_test.go (hunk 2 above)
Comment (b00ris) [reply e2e-polybft/txpool_test.go:165]: It make sense to have a special ticket for this problem. The best solution for me will be to have stable testutil.Contract that can generate any testing contracts. So, I propose to leave this code as it is, create an issue for testing smart contract storage and refactor this code as a part of that issue.

**Review by ferranbt — APPROVED** [review]

**Review by goran-ethernal — APPROVED** [review]

File: e2e-polybft/txpool_test.go (hunk 2 above)
Comment (igorcrevar) [reply e2e-polybft/txpool_test.go:165]: Manage to make it work with `testutil`. Only downside is that `block.chainid` op is removed because otherwise compiling with `testutil` fails(probably because older solidity version). https://github.com/0xPolygon/polygon-edge/pull/1140/commits/e2ffa832a2160c78ff99762ec13729eb2e9e4d1b

File: e2e-polybft/txpool_test.go (hunk 1 above)
Comment (igorcrevar) [reply e2e-polybft/txpool_test.go]: fixed https://github.com/0xPolygon/polygon-edge/pull/1140/commits/e2ffa832a2160c78ff99762ec13729eb2e9e4d1b

File: e2e-polybft/txpool_test.go (hunk 2 above)
Comment (igorcrevar) [reply e2e-polybft/txpool_test.go:165]: On my local machine compiling was working with previously mentioned commit but on github it does not. I am reverting those changes and we will fix that in another pr

File: e2e-polybft/txpool_test.go [hunk 3]
```diff
@@ -159,6 +161,36 @@ func TestE2E_TxPool_Transfer_Linear(t *testing.T) {
...
+
+	customArtifact, err := artifact.DecodeArtifact([]byte(smJSON))
+	require.NoError(t, err)
+
+	sidechainKey, err := wallet.GenerateKey()
+	require.NoError(t, err)
+
+	cluster := framework.NewTestCluster(t, 4,
+		framework.WithPremine(types.Address(sidechainKey.Address())),
```
Comment (Stefan-Ethernal) [inline e2e-polybft/txpool_test.go:174]: Does it make sense to shorten the epoch? Although this is not related to waiting epoch ending blocks. Anyhow if you feel it would be beneficial to shorten epoch, you can use `framework.WithEpochSize()` function.

File: e2e-polybft/txpool_test.go [hunk 4]
```diff
@@ -159,6 +161,36 @@ func TestE2E_TxPool_Transfer_Linear(t *testing.T) {
...
+
+	receipt, err := relayer.SendTransaction(&ethgo.Transaction{Input: customArtifact.Bytecode}, sidechainKey)
+	require.NoError(t, err)
+	require.Equal(t, uint64(types.ReceiptSuccess), receipt.Status)
+
+	receipt, err = ABITransaction(relayer, sidechainKey, customArtifact, receipt.ContractAddress, "init", []interface{}{})
+	require.NoError(t, err)
+	require.Equal(t, uint64(types.ReceiptSuccess), receipt.Status)
+
+	require.NoError(t, cluster.WaitForBlock(25, 2*time.Minute))
```
Comment (Stefan-Ethernal) [inline e2e-polybft/txpool_test.go]: Maybe you don't need to wait so long here? 

I'd say it would be enough to:
1. shorten epoch size to let's say 5 blocks
2. wait for block 10

File: e2e-polybft/txpool_test.go [hunk 5]
```diff
@@ -159,6 +161,36 @@ func TestE2E_TxPool_Transfer_Linear(t *testing.T) {
...
+
+	sidechainKey, err := wallet.GenerateKey()
+	require.NoError(t, err)
+
+	cluster := framework.NewTestCluster(t, 4,
+		framework.WithPremine(types.Address(sidechainKey.Address())),
+	)
+	defer cluster.Stop()
+
+	require.NoError(t, cluster.WaitForBlock(5, 30*time.Second))
```
Comment (Stefan-Ethernal) [inline e2e-polybft/txpool_test.go]: Is it enough here to wait for just a single block?

**Review by Stefan-Ethernal — APPROVED** [review]
Generally looks good, leaving some small suggestions to shorten the test.

File: e2e-polybft/txpool_test.go (hunk 5 above)
Comment (igorcrevar) [reply e2e-polybft/txpool_test.go]: https://github.com/0xPolygon/polygon-edge/pull/1140/commits/cb1fa0a0415784fa391b7dd3961651021f9d934a

File: e2e-polybft/txpool_test.go (hunk 4 above)
Comment (igorcrevar) [reply e2e-polybft/txpool_test.go]: size of epoch really does not matters for a test, but changed to wait for 10  https://github.com/0xPolygon/polygon-edge/pull/1140/commits/cb1fa0a0415784fa391b7dd3961651021f9d934a

File: e2e-polybft/txpool_test.go (hunk 3 above)
Comment (igorcrevar) [reply e2e-polybft/txpool_test.go:174]: The size of epoch does not really matter for the test
